Mods/ArchipelagoPlayer.SC2Mod/Base.SC2Data/Lib15EF4C78.galaxy

```diff
@@ -2874,6 +2874,190 @@ int lib15EF4C78_gf_AP_Player_getKerriganLevel (int lp_player) {
+            continue;
+        }
+
+        if (((UnitGetType(autoA0E4A284_var) == "AP_Extractor") || (UnitGetType(autoA0E4A284_var) == "AP_ExtractorPipes") || (UnitGetType(autoA0E4A284_var) == "AP_AutomatedExtractor") || (UnitGetType(autoA0E4A284_var) == "AP_Refinery") || (UnitGetType(autoA0E4A284_var) == "AP_AutomatedRefinery") || (UnitGetType(autoA0E4A284_var) == "AP_Assimilator") || (UnitGetType(autoA0E4A284_var) == "AP_AssimilatorProtossGeyser") || (UnitGetType(autoA0E4A284_var) == "AP_AssimilatorPurifierGeyser") || (UnitGetType(autoA0E4A284_var) == "AP_AssimilatorShakurasGeyser") || (UnitGetType(autoA0E4A284_var) == "AP_AutomatedAssimilator"))) {
+            lib15EF4C78_gf_ReplaceGasStructure(autoA0E4A284_var, lp_desiredFaction);
+            continue;
+        }
+
+        if ((CatalogFieldValueGet(c_gameCatalogUnit, UnitGetType(autoA0E4A284_var), "Race", c_playerAny) != CatalogFieldValueGet(c_gameCatalogRace, lp_desiredFaction, "AttributeId", c_playerAny))) {
+            UnitRemove(autoA0E4A284_var);
+        }
+
+    }
+}
+
+void lib15EF4C78_gf_ReplaceGasStructure (unit lp_gasStructure, string lp_desiredFaction) {
+    // Variable Declarations
+    int lv_initialGas;
+
+    // Automatic Variable Declarations
+    string auto148E77B6_val;
+    unitgroup autoEA05D24C_g;
+    int autoEA05D24C_u;
+    unit autoEA05D24C_var;
+
+    // Variable Initialization
+
+    // Implementation
+    lv_initialGas = UnitGetPropertyInt(lp_gasStructure, c_unitPropResources, c_unitPropCurrent);
+    if ((lv_initialGas > 0)) {
+        auto148E77B6_val = lp_desiredFaction;
+        if (auto148E77B6_val == "Terr") {
+            if ((UnitGetType(lp_gasStructure) != "AP_Refinery")) {
+                libNtve_gf_ReplaceUnit(lp_gasStructure, "AP_Refinery", libNtve_ge_ReplaceUnitOptions_NewUnitsDefault);
+                UnitSetPropertyFixed(libNtve_gf_LastReplacedUnit(), c_unitPropResources, lv_initialGas);
+            }
+
+        }
+        else if (auto148E77B6_val == "Zerg") {
+            if ((UnitGetType(lp_gasStructure) != "AP_Extractor") && (UnitGetType(lp_gasStructure) != "AP_ExtractorPipes")) {
+                libNtve_gf_ReplaceUnit(lp_gasStructure, "AP_Extractor", libNtve_ge_ReplaceUnitOptions_NewUnitsDefault);
+                UnitSetPropertyFixed(libNtve_gf_LastReplacedUnit(), c_unitPropResources, lv_initialGas);
+            }
+
+        }
+        else if (auto148E77B6_val == "Prot") {
+            if ((UnitGetType(lp_gasStructure) != "AP_Assimilator") && (UnitGetType(lp_gasStructure) != "AP_AssimilatorProtossGeyser") && (UnitGetType(lp_gasStructure) != "AP_AssimilatorPurifierGeyser") && (UnitGetType(lp_gasStructure) != "AP_AssimilatorShakurasGeyser")) {
+                libNtve_gf_ReplaceUnit(lp_gasStructure, "AP_Assimilator", libNtve_ge_ReplaceUnitOptions_NewUnitsDefault);
+                UnitSetPropertyFixed(libNtve_gf_LastReplacedUnit(), c_unitPropResources, lv_initialGas);
+            }
+

```

Comment:
I'd also check against the automated/rich variants in order to avoid race conditions/possible bugs

Reply:
Given that rich variants don't exist in the base game, I didn't handle them at all, particularly because they'd need to be handled differently. 

Agreed on autogas though, I've added those in.